src/ext/database_diagram.py

```diff
@@ -0,0 +1,188 @@
+    ]
+
+
+def get_columns_metadata(table_name, conn):
+    """Return the metadata for the columns in a table.
+
+    Args:
+        table_name (str): name of the DB table to retrieve the columns.
+        conn (sqlite3.Connection): SQLite connection.
+    Returns:
+        list: an array with the cid, name, type, notnull, default_value,
+        and pk value (0=not, 1=yes).
+    """
+    with closing(conn.cursor()) as cursor:
+        cursor.execute(f'PRAGMA table_info({table_name})')
+        return cursor.fetchall()
+
+
+def schema_to_markdown(db_name):
+    """Return the database markdown schema.
+
+    Args:
+        db_name (str): database name
+    Returns:
+        tuple: one side with lines of a markdown file, the other with
+            the orphan tables
+    """
+    lines = []
+    orphans = []
+    relationships = get_relationships()
+    with closing(sqlite3.connect(db_name)) as conn:
+        with closing(conn.cursor()) as cursor:
+            cursor.execute("SELECT name FROM sqlite_master WHERE type='table'")
+            for row in cursor:
+                table = row[0]
+                lines.append(f"[{table}]")
+                for column in get_columns_metadata(table, conn):
+                    pk = "*" if column[5] > 0 else ""  # 1 is normally PK
+                    label = f"{{label: {column[2].upper()}}}"
+                    line = f"\t{pk}{column[1]} {label}"
+                    lines.append(line)
+
+                if not any((
+                    table in relationship
+                    for relationship in relationships
+                )):
+                    orphans.append(table)
+
+            # in the eralchemy example, relationships go at the end of the file
+            lines.extend([" ".join(value) for value in relationships])
+            lines.extend(orphans)
+    return lines, orphans
+
+
+def group_nodes(nodes):
+    """Place the provided nodes in a subgraph to keep them together."""
+    lines = [
+        'subgraph {',
+        'rank=same',
+        'rankdir=LR'
+    ]
+    lines.extend([f'"{x}"' for x in nodes])
+    return lines + ['}']
+
+
+class CylcRunDBDirective(Graphviz):
+    """Directive which inserts a Cylc run DB graph."""
+
+    option_spec = dict(Graphviz.option_spec)
+    required_arguments = 0
+
+    @staticmethod
+    def generate_dotcode():
+        # get markdown
+        with tempfile.NamedTemporaryFile() as tf_db:
+            # is_public=False triggers the creation of tables
+            CylcSuiteDAO(db_file_name=tf_db.name, is_public=False)
+            schema, orphans = schema_to_markdown(db_name=tf_db.name)
+
+        # graph prefix
+        dotcode = [
+            'graph {',
+            'node [label = "\\N", shape = plaintext];',
+            'edge [color = gray50, minlen = 2, style = dashed];',
+            'rankdir = "TB";'
+        ]
+
+        # the database graph
+        tables, relationships = all_to_intermediary(schema)
```

Comment:
BTW I've dug into the `eralchemy` API to avoid having to work with files.

Reply:
Excellent! Thousand times better!